src/map.cpp

```diff
@@ -3017,315 +3017,300 @@ ter_id map::get_roof( const tripoint &p, const bool allow_air )
+            if( p == t || !has_flag( "COLLAPSES", t ) ) {
+                continue;
             }
-        // Made furniture seperate from the other tent to facilitate destruction
-        } else if (furnid == f_center_groundsheet || furnid == f_large_groundsheet ||
-                 furnid == f_large_canvas_door || furnid == f_large_canvas_wall ||
-                 furnid == f_large_canvas_door_o) {
-            if (str >= rng(0, 6) || destroy) {
-                // Special code to collapse the tent if destroyed
-                tripoint tentp = tripoint_min;
-                // Find the center of the tent
-                tripoint tmp = p;
-                int &i = tmp.x;
-                int &j = tmp.y;
-                for( i = p.x - 2; i <= p.x + 2; i++ ) {
-                    for( j = p.y - 2; j <= p.y + 2; j++ ) {
-                        if( furn(tmp) == f_center_groundsheet ){
-                            tentp = tmp;
-                            break;
-                        }
-                    }
-                }
-                // Never found tent center, bail out
-                if( tentp == tripoint_min ) {
-                    return std::pair<bool, bool>( true, false );
-                }
-                // Take the tent down
-                for( i = tentp.x-1; i <= tentp.x+1; i++ ) {
-                    for( j = tentp.y-1; j <= tentp.y+1; j++ ) {
-                         if (furn(tmp) == f_center_groundsheet) {
-                         spawn_item(tmp, "largebroketent");
-                        }
-                        furn_set(tmp, f_null);
-                    }
-                }
-                sound_volume = 8;
-                sound = _("rrrrip!");
-                smashed_something = true;
-                success = true;
-            } else {
-                sound_volume = 8;
-                sound = _("slap!");
-                smashed_something = true;
+
+            if( one_in( collapse_check( t ) ) ) {
+                collapse_at( t );
             }
         }
     }
 
-    if( move_cost(p) <= 0  && !smashed_something ) {
-        sound = _("thump!");
-        sound_volume = 18;
-        smashed_something = true;
-    }
-    if( !sound.empty() && !silent ) {
+    params.did_bash = true;
+    params.success |= success; // Not always true, so that we can tell when to stop destroying
+    params.bashed_solid = true;
+    if( !sound.empty() && !params.silent ) {
         sounds::sound( p, sound_volume, sound, false, "bash", sound );
     }
+}
+
+bash_params map::bash( const tripoint &p, const int str,
+                       bool silent, bool destroy, bool bash_floor,
+                       const vehicle *bashing_vehicle )
+{
+    bash_params bsh{
+        str, silent, destroy, bash_floor, (float)rng_float( 0, 1.0f ), false, false, false
+    };
+
+    // Remove webs
+    bash_field( p, bsh );
+
+    // Bash glass items
+    bash_items( p, bsh );
+
+    // Bash vehicle
```

Comment:
Sorry for the rant.

The function is named `bash_vehicle` and the comment says "bash vehicle" - what a surprising and useful comment. I had guessed that `bash_vehicle` is used to format the hard drive. Good thing the comment informed me what it actual does. /irony

Similar the comments above contain things about the implementation of the called function. `bash_field` does remove webs, but what if someone changes it to also remove other fields? Now the comment here is at least incomplete. Similar with `bash_items`, what if someone adds code to damage other items as well? (Ceramic may be fragile, plastic can also be bashed.)

In short: I think the function names speak for themselves and don't need to be commented here.